src/basic/terminal-util.c

```diff
@@ -1842,3 +1842,224 @@ int get_default_background_color(double *ret_red, double *ret_green, double *ret
+                                r = -EOPNOTSUPP;
+                                goto finish;
+                        }
+
+                        ssize_t l = read(input_fd, buf, sizeof(buf) - buf_full);
+                        if (l < 0) {
+                                r = -errno;
+                                goto finish;
+                        }
+
+                        assert((size_t) l <= sizeof(buf) - buf_full);
+                        buf_full += l;
+                }
+
+                size_t processed;
+                r = scan_cursor_position_response(&context, buf, buf_full, &processed);
+                if (r < 0)
+                        goto finish;
+
+                assert(processed <= buf_full);
+                buf_full -= processed;
+                memmove(buf, buf + processed, buf_full);
+
+                if (r > 0) {
+                        if (saved_row == 0) {
+                                assert(saved_column == 0);
+
+                                /* First sequence, this is the cursor position before we set it somewhere
+                                 * into the void at the bottom right. Let's save where we are so that we can
+                                 * return later. */
+
+                                /* Superficial validity checks */
+                                if (context.row <= 0 || context.column <= 0 || context.row >= 32766 || context.column >= 32766) {
+                                        r = -ENODATA;
+                                        goto finish;
+                                }
+
+                                saved_row = context.row;
+                                saved_column = context.column;
+
+                                /* Reset state */
+                                context = (CursorPositionContext) {};
+                        } else {
+                                /* Second sequence, this is the cursor position after we set it somewhere
+                                 * into the void at the bottom right. */
+
+                                /* Superficial validity checks */
+                                if (context.row < 4 || context.column < 4 || context.row >= 32766 || context.column >= 32766) {
+                                        r = -ENODATA;
+                                        goto finish;
+                                }
+
+                                if (ret_rows)
+                                        *ret_rows = context.row;
+                                if (ret_columns)
+                                        *ret_columns = context.column;
+
+                                r = 0;
+                                goto finish;
+                        }
+                }
+        }
+
+finish:
+        /* Restore cursor position */
+        if (saved_row > 0 && saved_column > 0)
+                RET_GATHER(r, terminal_set_cursor_position(output_fd, saved_row, saved_column));
```

Comment:
I wonder if we should fall back to the ioctl if we didn't manage to read the size from the terminal?

Reply:
hmm? not sure i follow? I think there's a misunderstanding here.

so the tty driver tracks terminal size in "struct winsize", which is readable by userspace via TIOCGWINSZ. On VTs the size is automatically managed by the VT driver. On ptys the size is automatically managed by whoever creates the pty (e.g. gnome-terminal). But on serial devices nothing configures the window size, which makes it default to 80x24. Which is almost always wrong.

This patch now adds support for asking a terminal connected via serial for its dimensions via ANSI sequences, and then updates from userspace the window size via TIOCSWINSZ. Thus from that point on we have a valid window size configured.

The whole point of terminal_get_size_by_dsr() is to do the ANSI sqeuence dance, and only that. Only if it succeeds, we have data to write using TIOCSWINSZ. Hence, it makes little sense to fallback to TIOCGWINSZ here if that's what you are suggesting, becasue that would mean we just write back the same incorrect data that is already set, which is pointless...